### PR TITLE
Document JUnit XML test results configuration options in user guide

### PR DESCRIPTION
**Pull request by ldaley**

### CODE REVIEW

**Comment by ldaley** [comment]
@bot-gradle test RFM

**Comment by ldaley** [comment]
@melix I requested your review as you're one of the best at writing useful docs. If you don't have the bandwidth, no problem.

**Comment by ldaley** [comment]
@bot-gradle test RFM

File: subprojects/docs/src/docs/release/notes.md [hunk 1]
```diff
@@ -248,37 +248,25 @@ same way it works for other Gradle API classes.
 
 ### Test re-run JUnit XML reporting enhancements
 
-The [`Test` task](dsl/org.gradle.api.tasks.testing.Test.html), used for [executing JVM tests](userguide/java_testing.html#test_reporting), reports test results as HTML and as a set of XML files in the “JUnit XML” pseudo standard.
+The `Test` task, used for executing JVM tests, reports test results as HTML and as a set of XML files in the “JUnit XML” pseudo standard.
 It is common for CI servers and other tooling to observe test results via the XML files.
-A new [`mergeReruns` option](javadoc/org/gradle/api/tasks/testing/JUnitXmlReport.html#getMergeReruns--) has been added that changes how tests that are executed more than once are reported in the XML files.
+A new `mergeReruns` has been added that changes how tests that are executed more than once are reported in the XML files.
```
Comment (eskatos) [inline subprojects/docs/src/docs/release/notes.md:253]:
```suggestion
A new `mergeReruns` option has been added that changes how tests that are executed more than once are reported in the XML files.
```

File: subprojects/docs/src/docs/release/notes.md [hunk 2]
```diff
@@ -248,37 +248,25 @@ same way it works for other Gradle API classes.
...
-With `mergeReruns` enabled, tests that pass-on-retry will no longer cause this Jenkins plugin to consider the build to have failed.
-However, failed test executions will be omitted from the Jenkins test result visualizations as it does not consider `<flakyFailure>` information.   
-The separate [Flaky Test Handler Jenkins plugin](https://plugins.jenkins.io/flaky-test-handler) can be used in addition to the JUnit Jenkins plugin to have
-such “flaky failures” also be visualized.
- 
-Note that enabling the new `mergeReruns` option does not add any retry/rerun functionality to test execution.
-Rerunning can be enabled by the test execution framework (e.g. JUnit's [`@RepeatedTest`](https://junit.org/junit5/docs/current/user-guide/#writing-tests-repeated-tests)), 
-or generically via the separate [Test Retry Gradle plugin](https://github.com/gradle/test-retry-gradle-plugin).
+
+This option is disable by default, causing each test execution to be listed as a separate `<testcase>` in the XML.
```
Comment (eskatos) [inline subprojects/docs/src/docs/release/notes.md:265]:
```suggestion
This option is disabled by default, causing each test execution to be listed as a separate `<testcase>` in the XML.
```

File: subprojects/docs/src/docs/userguide/jvm/java_testing.adoc [hunk 3]
```diff
@@ -234,6 +234,77 @@ Lastly, we add a `testReport` task that aggregates the test results from the `te
...
+
+===== mergeReruns
+
+When `mergeReruns` is enabled, if a test fails but is then retried and succeeds, its failures will be recorded as `<flakyFailure>` instead of `<failure>`, within one `<testcase>`.
+This is effectively the reporting produced by the link:https://maven.apache.org/components/surefire/maven-surefire-plugin/examples/rerun-failing-tests.html[surefire plugin of Apache Maven™] when enabling reruns.
+If your CI server understands this format, it will indicate that the test was flaky.
+If it does not, it will indicate that the test succeeded as it will ignore the `<flakyFailure>` information.
+If the test does not succeed (i.e. it fails for every retry), it will be indicated as having failed whether your tool understands this format or not.
+
+When `mergeReruns` is disabled (the default), each execution of a test be listed as an separate test case.
```
Comment (eskatos) [inline subprojects/docs/src/docs/userguide/jvm/java_testing.adoc:291]:
```suggestion
When `mergeReruns` is disabled (the default), each execution of a test will be listed as a separate test case.
```

File: subprojects/docs/src/docs/userguide/jvm/java_testing.adoc [hunk 4]
```diff
@@ -234,6 +234,77 @@ Lastly, we add a `testReport` task that aggregates the test results from the `te
...
+Enabling this option is especially useful when using a CI tool that uses the XML test results to determine build failure instead of relying on Gradle's determination of whether the build failed or not,
+and you wish to not consider the build failed if all failed tests passed when retried.
+This is the case for the Jenkins CI server and its link:https://plugins.jenkins.io/junit/[JUnit plugin].
+With `mergeReruns` enabled, tests that pass-on-retry will no longer cause this Jenkins plugin to consider the build to have failed.
+However, failed test executions will be omitted from the Jenkins test result visualizations as it does not consider `<flakyFailure>` information.
+The separate link:https://plugins.jenkins.io/flaky-test-handler[Flaky Test Handler Jenkins plugin] can be used in addition to the JUnit Jenkins plugin to have such “flaky failures” also be visualized.
+
+Tests are grouped and merged based on their reported name.
+When using any kind of test parameterization that affects the reported test name,
+or any other kind of mechanism the produces a potentially dynamic test name,
```
Comment (eskatos) [inline subprojects/docs/src/docs/userguide/jvm/java_testing.adoc:302]:
```suggestion
or any other kind of mechanism that produces a potentially dynamic test name,
```

File: subprojects/docs/src/docs/userguide/jvm/java_testing.adoc [hunk 5]
```diff
@@ -234,6 +234,77 @@ Lastly, we add a `testReport` task that aggregates the test results from the `te
...
+care should be take to ensure that the test name is stable and does not unnecessarily change.
```
Comment (eskatos) [inline subprojects/docs/src/docs/userguide/jvm/java_testing.adoc:303]:
```suggestion
care should be taken to ensure that the test name is stable and does not unnecessarily change.
```

File: subprojects/docs/src/snippets/testing/junit-xml-results/kotlin/build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+}
+// end::configure-location-task[]
+
+// tag::configure-location-convention[]
+project.setProperty("testResultsDirName", "$buildDir/junit-xml")
+// end::configure-location-convention[]
+
+// tag::configure-content[]
+tasks.test {
+    reports.apply {
```
Comment (eskatos) [inline subprojects/docs/src/snippets/testing/junit-xml-results/kotlin/build.gradle.kts:27]: This should work as `AbstractTestTask` has the corresponding `Action` taking method.
```suggestion
    reports {
```

**Review by eskatos — COMMENTED** [review]
Some text nit picks, and a suggestion to simplify a Kotlin DSL script snippet.

**Comment by eskatos** [comment]
@bot-gradle test this

**Review by eskatos — APPROVED** [review]

**Comment by melix** [comment]
Thanks for reviewing @eskatos!

**Comment by ldaley** [comment]
@eskatos thanks for merging.

**Pull request merged**